File: tools/score_licenses/score.php
<?php
require_once('config.php');

// load the configurations of rating weights for the different license aspects
$json = file_get_contents($WEIGHT_FILE);
$weights = json_decode($json); 
if(empty($weights)){
  print "Unable to read JSON weights file.\n";
  exit(1);
}

// calculate ratings for each license
$ratings = array();
foreach(list_licenses() as $id){
   print "Scoring license $id. ";
	$metadata = get_license_metadata($id);
	
	// calculate the rating for each category
	$ratings[$id]['freedom_score'] = rate_freedom($metadata, $weights->license_components);
	$ratings[$id]['legal_risk_score'] = rate_legal_risk($metadata, $weights->license_components);
	$ratings[$id]['business_risk_score'] = rate_business_risk($metadata, $weights->license_components);
	
	// calculate the overall rating
	$ratings[$id]['overall_score'] =
	  $ratings[$id]['freedom_score'] * $weights->categories->freedom +
	  $ratings[$id]['legal_risk_score'] * $weights->categories->legal_risk +
	  $ratings[$id]['business_risk_score'] * $weights->categories->business_risk;
	  
	print 'Overall score = ' . $ratings[$id]['overall_score'] . "\n";
}

// calculate ranking for each type of score
foreach(array('freedom', 'legal_risk', 'business_risk', 'overall') as $score_type){
	
	// get a list of the scores for this score type
	$scores = array();
	foreach($ratings as $key => $val)
	  $scores[] = $val[$score_type . '_score'];
	
	// sort by score and add ranking data
	array_multisort($scores, SORT_DESC, SORT_NUMERIC, $ratings);
	$rank = 1;
	foreach($ratings as $key => $val){
		// assign a rank
		$ratings[$key][$score_type . '_rank'] = $rank++;
		
		// now that ranking is complete, we can format the score per the configured output format
		$ratings[$key][$score_type . '_score'] =
		  sprintf($SCORE_FORMAT, $ratings[$key][$score_type . '_score']);
	}
}

// save the results
foreach($ratings as $id => $val){
	 // alphabetically sort the result data by key and convert to json
	 ksort($val);
	 $json = json_encode($val, JSON_PRETTY_PRINT);
	 
	 // write out the result
	 $filename = $LICENSE_SCORE_DIR . '/' . $id . '.json';
    if(($fh = fopen($filename, 'w')) === false){
	   print "Unable to write file ($filename).\n";
	   exit(1);
	 }
    fwrite($fh, $json);
    fclose($fh);
}

// assigns a score of the legal risk associated with a license (high score = low risk to the user)
function rate_legal_risk($license_data, $weights){

  // Choice of forum affects legal risks because of the increased legal costs of defending against a lawsuit in a foreign jurisdiction
  // -A forum selection clause (FSC) for the jurisdiction of the defendant generally introduces the lowest cost for the defendant, deterring lawsuits and legal risk
  // -An unspecified forum leaves the issue of forum up to the court to decide based on conflict of law rules, often involving considerable discretion such as with forum non conveniens considerations
  // -A FSC in favour of the licensor or pl will often bring the forum outside of the def's home jurisdictions (where this is different), introducing the highest degree of legal risk
  $forum_score = null;
  if($license_data->conflict_of_laws->forum_of == 'defendant')
    $forum_score = 1;
  else if($license_data->conflict_of_laws->forum_of == 'unspecified')
    $forum_score = 0.5;
  else if($license_data->conflict_of_laws->forum_of == 'licensor' ||
          $license_data->conflict_of_laws->forum_of == 'plaintiff' ||
          $license_data->conflict_of_laws->forum_of == 'specific')
    $forum_score = 0;
  else{
    print "Error: unknown value for conflict_of_law\n";
    exit(1);
  }
  
  // Choice of law affects legal risks because of the increased legal costs of retaining counsel and experts familiar with a foreign law
  // (it also introduces uncertainty of the legal risks in using a license where foreign legal rules may apply)
  // -A choice of law clause (CLC) for the jurisdiction of the defendant generally introduces the lowest cost for the defendant, deterring lawsuits and legal risk
  // -An unspecified CSC leaves the issue of forum up to the court to decide based on conflict of law rules, often involving considerable discretion such as with forum non conveniens considerations
  // -A CLC in favour of the licensor or pl will often bring the forum outside of the def's home jurisdictions (where this is different), introducing the highest degree of legal risk
  // -a CLC in favour of the forum which match the legal risk for the forum
  $col_score = null;
  if($license_data->conflict_of_laws->law_of == 'defendant')
    $col_score = 1;
  else if($license_data->conflict_of_laws->law_of == 'unspecified')
    $col_score = 0.5;
  else if($license_data->conflict_of_laws->law_of == 'licensor' ||
          $license_data->conflict_of_laws->law_of == 'plaintiff' ||
          $license_data->conflict_of_laws->law_of == 'specific')
    $col_score = 0;
  else if($license_data->conflict_of_laws->law_of == 'forum'){
  	 $col_score = $forum_score;
  }
  else{
    print "Error: unknown value for conflict_of_law in the following license data:\n";
    print_r($license_data);
    exit(1);
  }
  
  // The two key warranties that licensor's can provide to reduce the licensee's legal risk are 1. a warranty
  // related to quality and accuracy, such as fitness for a purpose or merchantability); and/or 2. a warranty of
  // noninfringement of copyright (and other IP)
  // -For 1, a form of this warranty is implied in many, if not most, jurisdictions.
  // -For 2, only some jurisdictions recognize an implied warranty of noninfringement.  Thus, this score is applied only
  // where such a warranty is explicitly provided and not merely where the license remains silent to this aspect. many other jurisdictions.
  $warranty_score = 0;
  if(!$license_data->disclaimers->disclaimer_warranty)
    $warranty_score += 0.5;
  if(!$license_data->disclaimers->warranty_noninfringement)
    $warranty_score += 0.5;

  // A disclaimer of liability generally prevents the licensee from making a claim against the licensor, even
  // where the licensor is legally at fault.
  $disclaimer_score = !$license_data->disclaimers->disclaimer_liability ? 1 : 0;

  // Indemnication clauses obligate the licensee to defend and/or compensate the licensor for any 3rd party
  // lawsuits against the licensor (even if the fault is attributable to the licensor)
  $indemnity_score = !$license_data->disclaimers->disclaimer_indemnity ? 1 : 0;     
    
  return
    $forum_score * $weights->choice_of_forum +
    $col_score * $weights->choice_of_law +
    $warranty_score * $weights->warranty +
    $disclaimer_score * $weights->disclaimer +
    $indemnity_score * $weights->indemnity;
}

// assigns a score of the business risk associated with a license (high score = low risk to the user)
function rate_business_risk($license_data, $weights){
  // The ability of the licensor to unilaterally change the license terms (effectively revoking the original license)
  // creates a risk that the new license terms will not be amenable to the planned or existing use case of a licensee
  $unilateral_changes = !$license_data->license_changes->license_changes_effective_immediately ? 1 : 0;
  
  // Where a license terminates on any violation of the terms, even an inadvertent breach could jeopordize
  // long-term use of the licensed content
  // -Where there is no automatic termination, damages or specific performance are generally the only remedies
  //  available (repudiation is only an option for fundamental breaches). This poses the lowest business risk.
  // -Where an automatic termination clause is attenuated by a chance to remedy a breach and automatically reinstate
  //  the license, inadvertent breaches pose less of a business risk.
  // -The right of the licensor to exercise discretion to revoke a license even where the licensee has not
  //  breached the terms poses the highest degree of business risk
  $termination = null;
  if($license_data->termination->termination_discretionary)
    $termination = 0;
  else if($license_data->termination->termination_automatic)
  	 $termination = $license_data->termination->termination_reinstatement ? 2.0/3.0 : 1.0/3.0;
  else
  	 $termination = 1.0;
  
  return
    $unilateral_changes *= $weights->unilateral_changes +
    $termination *= $weights->termination;
}

// assigns a score of the amount of freedom granted by a license (high score = more freedom);
// Note: does not deal with attempts by a license to maintain freedoms (ie. copyleft or restrictions on TPMs)
function rate_freedom($license_data, $weights){
   // With respect to data licenses, copyright is the only right that is likely to matter in most cases
   // (except in jurisdictions where SGDRs apply). This rights needs to be granted or the license
   // essentially grants nothing at all.
   if(!$license_data->rights->covers_copyright)
     return 0;
   	
	// The core rights that a license can grant and to use, modify and distribute
	$core_rights = 0;
	if($license_data->rights->right_to_use_and_reproduce)
	  $core_rights += 1.0/3.0;
	if($license_data->rights->right_to_modify)
	  $core_rights += 1.0/3.0;
	if($license_data->rights->right_to_distribute)
     $core_rights += 1.0/3.0;
     
   // The above rights may be restricted to non-commercial purposes\
   $commercial_use = !$license_data->rights->prohibits_commercial_use ? 1.0 : 0;
   
   // Notice and attribution requirements can reduce user freedom due to the burdens that complying with these
   // obligations can impose (they can become particularly problematic when combining data or content from
   // many sources)
   // -Specific attribution imposes specific wording, which may not always be feasible depending on the
   //  format (for example, a license could not necessarily comply with attribution in metadata of a large dataset)
   // -Flexible attribution generally allows a licensee to attribute (and provide notice if this is a requirement)
   //  in a manner appropriate to the medium 
   // -A mere notice requirements usually only requires a reasonably accessible URL or other notice to indicate
   //  the license that the material falls under 
  $notice_and_attribution = null;
  if($license_data->obligations->obligation_attribution_specific)
    $notice_and_attribution = 0;
  else if($license_data->obligations->obligation_attribution_flexible)
    $notice_and_attribution = 1.0 / 3.0;
  else if($license_data->obligations->obligation_notice)
    $notice_and_attribution = 2.0 / 3.0;
  else
    $notice_and_attribution = 1.0;
    
  return
    $core_rights * $weights->core_rights +
    $commercial_use * $weights->commercial_use +
    $notice_and_attribution * $weights->attribution;
}

// load the license metadata for a particular license
function get_license_metadata($id){
  global $LICENSE_DATA_DIR;

  // read the json and convert to a php object
  $json_filename = $LICENSE_DATA_DIR . '/' . $id . '.json';
  $json = file_get_contents($json_filename);
  return json_decode($json); 
  
}

// get a list of the licenses in the license dir
function list_licenses(){
  global $LICENSE_DATA_DIR;

  $ids = array();
  if($dirHandle = opendir($LICENSE_DATA_DIR)){
    while(false !== ($entry = readdir($dirHandle))){
      if(is_file($LICENSE_DATA_DIR . '/' . $entry)){
        $ids[] = substr($entry,0,-5);
      }
    }
  }
  sort($ids);
  return $ids;
}
	
?>

File: app/models/score.rb
class Score < ActiveRecord::Base
  belongs_to :licence
  attr_accessible :openness, :licensee_legal_risk, :licensee_business_risk, :licensee_freedom, :licensor_legal_risk

  SCORE_WEIGHTS = {
      categories: {
          freedom: 1.0,
          legal_risk: 1.0,
          business_risk: 1.0
      },
      attributes: {
          core_rights: 13.0, # everything else only applies insofar as the core rights grant freedoms
          commercial_use: 1.0,
          attribution: 1.0,

          choice_of_forum: 0.75,
          choice_of_law: 0.75,
          warranty: 0.75,
          disclaimer: 0.75,
          indemnity: 2.0,

          unilateral_changes: 2.0,
          termination: 3.0
      }
  }

  # calculate the scores
  def score_licence
    calculate_licensee_legal_risk
    calculate_licensee_business_risk
    calculate_licensee_freedom
    calculate_licensor_legal_risk
    calculate_overall_openness
  end

  # >=22.5 = 3 stars (Gold), >=20 = 2 stars (Silver), >= 17.5 = 1 star (Bronze)
  def star_rating
    case
      when openness > 22.5
        3
      when openness > 20
        2
      when openness > 17.5
        1
      else
        0
    end
  end

  def openness_rank
    Score.count(conditions: "openness > #{self.openness}" ) + 1
  end

  def licensee_freedom_rank
    Score.count(conditions: "licensee_freedom > #{self.licensee_freedom}" ) + 1
  end

  def licensee_legal_risk_rank
    Score.count(conditions: "licensee_legal_risk > #{self.licensee_legal_risk}" ) + 1
  end

  def licensee_business_risk_rank
    Score.count(conditions: "licensee_business_risk > #{self.licensee_business_risk}" ) + 1
  end

  def licensor_business_risk_rank
    Score.count(conditions: "licensor_business_risk > #{self.licensor_business_risk}" ) + 1
  end

  private

  # overall measure of the "openness" of the licence (rather subjective, but suggestions for improvements are welcome!)
  def calculate_overall_openness
    self.openness = self.licensee_legal_risk    * SCORE_WEIGHTS[:categories][:legal_risk] +
        self.licensee_business_risk * SCORE_WEIGHTS[:categories][:business_risk] +
        self.licensee_freedom       * SCORE_WEIGHTS[:categories][:freedom]
  end

  # calculate a score of the legal risk to the licensee
  def calculate_licensee_legal_risk

    # Choice of forum affects legal risks because of the increased legal costs of defending against a lawsuit
    # in a foreign jurisdiction
    # - A forum selection clause (FSC) for the jurisdiction of the defendant generally introduces the lowest
    #   cost for the defendant, deterring lawsuits and legal risk
    # - An unspecified forum leaves the issue of forum up to the court to decide based on conflict of law
    #   rules, often involving considerable discretion such as with forum non conveniens considerations
    # - A FSC in favour of the licensor or pl will often bring the forum outside of the def's home
    #   jurisdictions (where this is different), introducing the highest degree of legal risk
    forum_score = nil
    if licence.conflict_of_law.forum_of == 'defendant'
      forum_score = 1.0
    elsif licence.conflict_of_law.forum_of == 'unspecified'
      forum_score = 0.5
    elsif licence.conflict_of_law.forum_of == 'licensor' ||
        licence.conflict_of_law.forum_of  == 'plaintiff' ||
        licence.conflict_of_law.forum_of  == 'specific'
      forum_score = 0
    else
      raise "Unrecognized value for conflict_of_law.forum_of"
    end

    # Choice of law affects legal risks because of the increased legal costs of retaining counsel
    # and experts familiar with a foreign law (it also introduces uncertainty of the legal risks in
    # using a license where foreign legal rules may apply)
    # - A choice of law clause (CLC) for the jurisdiction of the defendant generally introduces the lowest
    #   cost for the defendant, deterring lawsuits and legal risk
    # - An unspecified CSC leaves the issue of forum up to the court to decide based on conflict of law
    #   rules, often involving considerable discretion such as with forum non conveniens considerations
    # - A CLC in favour of the licensor or pl will often bring the forum outside of the def's home
    #   jurisdictions (where this is different), introducing the highest degree of legal risk
    # - A CLC in favour of the forum will match the legal risk for the forum
    law_score = nil
    if licence.conflict_of_law.law_of == 'defendant'
      law_score = 1.0
    elsif licence.conflict_of_law.law_of == 'unspecified'
      law_score = 0.5
    elsif licence.conflict_of_law.law_of == 'licensor' ||
        licence.conflict_of_law.law_of == 'plaintiff' ||
        licence.conflict_of_law.law_of == 'specific'
      law_score = 0
    elsif licence.conflict_of_law.law_of == 'forum'
      law_score = forum_score
    else
      raise "Unrecognized value for conflict_of_law.law_of"
    end

    # The two key warranties that licensor's can provide to reduce the licensee's legal risk are
    # 1. a warranty related to quality and accuracy, such as fitness for a purpose or merchantability; and/or
    # 2. a warranty of noninfringement of copyright (and other IP)
    # For #1, a form of this warranty is implied in many, if not most, jurisdictions. A disclaimer will negate it.
    # For #2, only some jurisdictions recognize an implied warranty of noninfringement.  Thus, a score for this is
    # applied only where such a warranty is explicitly provided.
    warranty_score = 0
    warranty_score += 0.5 if !licence.disclaimer.disclaimer_warranty
    warranty_score += 0.5 if licence.disclaimer.warranty_noninfringement

    # A disclaimer of liability generally prevents the licensee from making a claim against the licensor, even
    # where the licensor is legally at fault.
    disclaimer_score = licence.disclaimer.disclaimer_liability ? 0.0 : 1.0

    # Indemnication clauses obligate the licensee to defend and/or compensate the licensor for any 3rd party
    # lawsuits against the licensor (even if the fault is attributable to the licensor)
    indemnity_score = licence.disclaimer.disclaimer_indemnity ? 0.0 : 1.0

    # total score for legal risk to the licence user
    self.licensee_legal_risk =
        forum_score      * SCORE_WEIGHTS[:attributes][:choice_of_forum] +
            law_score        * SCORE_WEIGHTS[:attributes][:choice_of_law] +
            warranty_score   * SCORE_WEIGHTS[:attributes][:warranty] +
            disclaimer_score * SCORE_WEIGHTS[:attributes][:disclaimer] +
            indemnity_score  * SCORE_WEIGHTS[:attributes][:indemnity]
  end

  # calculate a score of the legal risk to the licensor
  def calculate_licensor_legal_risk
    # legal risk to the licencensor is the inverse of the risk to the licencee
    max_licensee_legal_risk =  SCORE_WEIGHTS[:attributes][:choice_of_forum] +
        SCORE_WEIGHTS[:attributes][:warranty] +
        SCORE_WEIGHTS[:attributes][:disclaimer] +
        SCORE_WEIGHTS[:attributes][:indemnity]

    self.licensor_legal_risk = max_licensee_legal_risk - self.licensee_legal_risk
  end

  # calculate a score of the business risk associated with a license (high score = low risk to the user)
  def calculate_licensee_business_risk
    # The ability of the licensor to unilaterally change the license terms (effectively revoking the original license)
    # creates a risk that the new license terms will not be amenable to the planned or existing business use case of a
    # licensee
    unilateral_change_score = licence.changes_to_term.licence_changes_effective_immediately ? 0 : 1.0

    # Where a license terminates on any violation of the terms, even an inadvertent breach could jeopordize
    # long-term use of the licensed content
    # Where there is no automatic termination, damages or specific performance are generally the only remedies
    # available (repudiation is only an option for fundamental breaches): this poses a lower business risk.
    # As well, where an automatic termination clause is attenuated by a chance to remedy a breach and automatically
    # reinstate the license, inadvertent breaches poses less of a business risk.
    # The right of the licensor to exercise discretion to revoke a license even where the licensee has not
    #  breached the terms poses the highest degree of business risk

    termination_risk_score = nil
    if licence.termination.termination_discretionary
      termination_risk_score = 0
    elsif licence.termination.termination_automatic
      termination_risk_score = 1.0 / 3.0
    else
      termination_risk_score = 1.0
    end

    termination_risk_score += (1.0 / 3.0) if licence.termination.termination_reinstatement

    self.licensee_business_risk =
        unilateral_change_score * SCORE_WEIGHTS[:attributes][:unilateral_changes] +
            termination_risk_score  * SCORE_WEIGHTS[:attributes][:termination]
  end

  # calculate a score of the degree of freedom that a licence grants (high score = more freedom);
  # (note: does not deal with attempts by a license to maintain freedoms (ie. copyleft or restrictions on TPMs)
  def calculate_licensee_freedom
    # With respect to data licenses, copyright is the only right that is likely to matter in most cases
    # (except in jurisdictions where SGDRs apply). This rights needs to be granted, or the license
    # effectively grants little
    if !licence.right.covers_copyright
      return (self.licensee_freedom = 0)
    end

    # the core rights that licenses typically grant and the right to use, modify & distribute
    core_rights_score = 0
    core_rights_score += (1.0/3.0) if licence.right.right_to_use_and_reproduce
    core_rights_score += (1.0/3.0) if licence.right.right_to_modify
    core_rights_score += (1.0/3.0) if licence.right.right_to_distribute

    # the above rights may be restricted to non-commercial purposes
    commercial_use_score = licence.right.prohibits_commercial_use ? 0 : 1.0

    # Notice and attribution requirements may reduce user freedom due to the burdens that complying with these
    # obligations can impose (they can become particularly problematic when combining data or content from
    # many sources)
    # Specific attribution imposes specific wording, which may not always be feasible depending on the
    # format (for example, a licensee could have difficulty complying with attribution using metadata for a large dataset)
    # Flexible attribution generally allows a licensee to attribute in a manner appropriate to the medium
    # A mere notice requirement usually only requires a reasonably accessible URL or other such notice to indicate
    # the particular licence under which the material falls
    notice_and_attribution_score = 1.0
    if licence.obligation.obligation_attribution
      notice_and_attribution_score -= (2.0 / 3.0)
      notice_and_attribution_score += (1.0 / 3.0) if licence.attribution_clause.attribution_type == 'flexible'
    end
    notice_and_attribution_score -= (1.0 / 3.0) if licence.obligation.obligation_notice

    self.licensee_freedom =
        core_rights_score            * SCORE_WEIGHTS[:attributes][:core_rights] +
            commercial_use_score         * SCORE_WEIGHTS[:attributes][:commercial_use] +
            notice_and_attribution_score * SCORE_WEIGHTS[:attributes][:attribution]
  end
end